lib/posix/pthread.c

```diff
@@ -410,54 +564,52 @@ void pthread_exit(void *retval)
 		}
 	}
 
-	pthread_mutex_unlock(&self->state_lock);
-	pthread_mutex_destroy(&self->state_lock);
-
-	pthread_cond_destroy(&self->state_cond);
-
-	k_thread_abort((k_tid_t)self);
+	posix_thread_finalize(self, retval);
+	CODE_UNREACHABLE;
 }
 
 /**
  * @brief Wait for a thread termination.
  *
  * See IEEE 1003.1
  */
-int pthread_join(pthread_t thread, void **status)
+int pthread_join(pthread_t th, void **status)
 {
-	struct posix_thread *pthread = to_posix_thread(thread);
-	int ret = 0;
+	int err;
+	int qid;
+	int detachstate;
+	k_spinlock_key_t key;
+	struct posix_thread *t;
 
-	if (thread == pthread_self()) {
+	t = to_posix_thread(th);
+	if (th == pthread_self()) {
 		return EDEADLK;
 	}
 
-	if (pthread == NULL) {
+	if (t == NULL) {
 		return ESRCH;
 	}
 
-	pthread_mutex_lock(&pthread->state_lock);
+	key = k_spin_lock(&t->lock);
+	qid = t->qid;
+	detachstate = t->detachstate;
+	k_spin_unlock(&t->lock, key);
 
-	if (pthread->state == PTHREAD_JOINABLE) {
-		pthread_cond_wait(&pthread->state_cond, &pthread->state_lock);
+	if (detachstate != PTHREAD_CREATE_JOINABLE) {
+		return EINVAL;
 	}
 
-	if (pthread->state == PTHREAD_EXITED) {
-		if (status != NULL) {
-			*status = pthread->retval;
-		}
-	} else if (pthread->state == PTHREAD_DETACHED) {
-		ret = EINVAL;
-	} else {
-		ret = ESRCH;
+	if (detachstate == PTHREAD_CREATE_JOINABLE && qid != POSIX_THREAD_READY_Q) {
```

Comment:
Hrm. Is there a possible race with fetching qid under the lock and testing it with the lock released?

Reply:
should be fixed now.